bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java

```diff
@@ -0,0 +1,656 @@
+            return refreshToken;
+        }
+        try {
+            final byte[] contents = Files.readAllBytes(Paths.get(fileName));
+            refreshToken = new String(contents);
+        } catch (IOException ex) {
+            logger.debug("IOException when reading refreshToken from file {}", ex.getMessage());
+        }
+        logger.debug("getRefreshTokenFromFile successful {}", RingUtils.sanitizeData(refreshToken));
+        return refreshToken;
+    }
+
+    public void doLogin(String username, String password, String twofactorCode) {
+        logger.debug("doLogin U:{} P:{} 2:{}", RingUtils.sanitizeData(username), RingUtils.sanitizeData(password),
+                RingUtils.sanitizeData(twofactorCode));
+        String hardwareId = getHardwareId();
+        String refreshToken = getRefreshTokenFromFile();
+        logger.debug("doLogin H:{} RT:{}", hardwareId, RingUtils.sanitizeData(refreshToken));
+        try {
+            userProfile = restClient.getAuthenticatedProfile(username, password, refreshToken, twofactorCode,
+                    hardwareId);
+            saveRefreshTokenToFile(userProfile.getRefreshToken());
+        } catch (AuthenticationException ex) {
+            logger.debug("AuthenticationException when initializing Ring Account handler{}", ex.getMessage());
+            if (ex.getMessage().startsWith("Two factor")) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, ex.getMessage());
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, ex.getMessage());
+            }
+        } catch (JsonParseException e) {
+            logger.debug("Invalid response from api.ring.com when initializing Ring Account handler{}", e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Invalid response from api.ring.com");
+        }
+        logger.debug("doLogin RT: {}", getRefreshTokenFromFile());
+        try {
+            refreshRegistry();
+            updateStatus(ThingStatus.ONLINE);
+        } catch (DuplicateIdException ignored) {
+            updateStatus(ThingStatus.ONLINE);
+        } catch (AuthenticationException ae) {
+            registry = null;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "AuthenticationException response from ring.com");
+            logger.debug("RestClient reported AuthenticationException in finally block: {}", ae.getMessage());
+        } catch (JsonParseException pe1) {
+            registry = null;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "JsonParseException response from ring.com");
+            logger.debug("RestClient reported JsonParseException in finally block: {}", pe1.getMessage());
+        }
+    }
+
+    public @NonNullByDefault({}) String getHardwareId() {
```

Comment:
why do you need this annotation?

Reply:
I don't :) I guess is was already there in the code I inherited, or I added it when trying to fix warnings. Either way, fixed now.